src/control/server/storage_scm.go

```diff
@@ -48,6 +67,166 @@ var (
 	msgScmUpdateNotImpl     = "scm firmware update not supported"
 )
 
+type runCmdFn func(string) ([]byte, error)
+
+// ScmSetup is an interface to configure scm prerequisites via shell tools
+type ScmSetup interface {
+	prep(common.ScmModules) ([]byte, error)
```

Comment:
Looking at this some more, I don't understand why everything has to return a byte slice in addition to an error. I think I understand that it comes from cmd.Output(), but it seems strange to make everything mirror that signature. As indicated in my other comment, it doesn't seem right to make the caller check the error and the returned byte slice to figure out what happened. Generally speaking you don't want library code to be concerned with UI (i.e. returning a string to maybe be displayed to a user).

Reply:
The other problem here is that the byte slice returned from cmd.Output() contains the command's stdout, but stderr is lost unless the caller knows to cast the error to an exec.ExitError and look at the Stderr field. I think it would be better to return a custom error implementation which includes stderr somehow in the Error() output. e.g.
```go
type scmError struct {
        wrapped error
}

func (se *scmError) Error() string {
        if ee, ok := se.wrapped.(*os.ExitError) {
                return fmt.Sprintf("%s: %s", ee.ProcessState, ee.Stderr)
        }
        return se.wrapped.Error()
}
```

Reply:
> Looking at this some more, I don't understand why everything has to return a byte slice in addition to an error. I think I understand that it comes from cmd.Output(), but it seems strange to make everything mirror that signature. As indicated in my other comment, it doesn't seem right to make the caller check the error and the returned byte slice to figure out what happened. Generally speaking you don't want library code to be concerned with UI (i.e. returning a string to maybe be displayed to a user).

this is a result of me trying to find a way of mocking external command output that meets your approval. previously I would have called into the external.go Interface runCommand() method which could then be provided as a mock implementation in the unit tests. seeing as you didn't like that approach I was trying to provide the mocking capability in a more localised fashion (hoping you would approve with that) by providing runCmdFn signature type that could be passed from caller and propagated to delegated methods and therefore mocked in test. 

The caller is expected to process either the error or the command output but not both.

Reply:
> The other problem here is that the byte slice returned from cmd.Output() contains the command's stdout, but stderr is lost unless the caller knows to cast the error to an exec.ExitError and look at the Stderr field. I think it would be better to return a custom error implementation which includes stderr somehow in the Error() output. e.g.
> 
> ```go
> type scmError struct {
>         wrapped error
> }
> 
> func (se *scmError) Error() string {
>         if ee, ok := se.wrapped.(*exec.ExitError); ok {
>                 return fmt.Sprintf("%s: %s", ee.ProcessState, ee.Stderr)
>         }
>         return se.wrapped.Error()
> }
> ```

yes I can do this